pkg/alibabacloud/eni/node.go

```diff
@@ -0,0 +1,396 @@
+					"availableAddresses": subnet.AvailableAddresses,
+				}).Debug("Subnet has IPs available")
+
+				a.InterfaceID = key
+				a.PoolID = ipamTypes.PoolID(subnet.ID)
+				a.AvailableForAllocation = math.IntMin(subnet.AvailableAddresses, availableOnENI)
+			}
+		}
+	}
+	a.AvailableInterfaces = l.Adapters - len(n.enis) + a.AvailableInterfaces
+	return a, nil
+}
+
+// AllocateIPs performs the ENI allocation operation
+func (n *Node) AllocateIPs(ctx context.Context, a *ipam.AllocationAction) error {
+	_, err := n.manager.api.AssignPrivateIpAddresses(ctx, a.InterfaceID, a.AvailableForAllocation)
+	return err
+}
+
+// PrepareIPRelease prepares the release of ENI IPs.
+func (n *Node) PrepareIPRelease(excessIPs int, scopedLog *logrus.Entry) *ipam.ReleaseAction {
+	r := &ipam.ReleaseAction{}
+
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+
+	// Iterate over ENIs on this node, select the ENI with the most
+	// addresses available for release
+	for key, e := range n.enis {
+		if e.Type != "Secondary" {
+			continue
+		}
+		scopedLog.WithFields(logrus.Fields{
+			fieldEniID:     e.NetworkInterfaceID,
+			"numAddresses": len(e.PrivateIPSets),
+		}).Debug("Considering ENI for IP release")
+
+		// Count free IP addresses on this ENI
+		ipsOnENI := n.k8sObj.Status.AlibabaCloud.ENIs[e.NetworkInterfaceID].PrivateIPSets
+		freeIpsOnENI := []string{}
+		for _, ip := range ipsOnENI {
+			// exclude primary IPs
+			if ip.Primary {
+				continue
+			}
+			_, ipUsed := n.k8sObj.Status.IPAM.Used[ip.PrivateIpAddress]
+			if !ipUsed {
+				freeIpsOnENI = append(freeIpsOnENI, ip.PrivateIpAddress)
+			}
+		}
+		freeOnENICount := len(freeIpsOnENI)
+
+		if freeOnENICount <= 0 {
+			continue
+		}
+
+		scopedLog.WithFields(logrus.Fields{
+			fieldEniID:       e.NetworkInterfaceID,
+			"excessIPs":      excessIPs,
+			"freeOnENICount": freeOnENICount,
+		}).Debug("ENI has unused IPs that can be released")
+		maxReleaseOnENI := math.IntMin(freeOnENICount, excessIPs)
+
+		r.InterfaceID = key
+		r.PoolID = ipamTypes.PoolID(e.Vpc.VpcID)
+		r.IPsToRelease = freeIpsOnENI[:maxReleaseOnENI]
+	}
+
+	return r
+}
+
+// ReleaseIPs performs the ENI IP release operation
+func (n *Node) ReleaseIPs(ctx context.Context, r *ipam.ReleaseAction) error {
+	return n.manager.api.UnassignPrivateIpAddresses(ctx, r.InterfaceID, r.IPsToRelease)
+}
+
+// GetMaximumAllocatableIPv4 returns the maximum amount of IPv4 addresses
+// that can be allocated to the instance
+func (n *Node) GetMaximumAllocatableIPv4() int {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+
+	// Retrieve l for the instance type
+	l, limitsAvailable := n.getLimitsLocked()
+	if !limitsAvailable {
+		return 0
+	}
+
+	// Return the maximum amount of IP addresses allocatable on the instance
+	// reserve Primary eni
+	return (l.Adapters - 1) * l.IPv4
+}
+
+func (n *Node) loggerLocked() *logrus.Entry {
+	if n == nil || n.node == nil {
+		return log
+	}
+
+	return log.WithField("instanceID", n.node.InstanceID())
+}
```

Comment:
This code will run into the same issues as seen with AWS ENI mode. Please see https://github.com/cilium/cilium/pull/14905 which fixes it and apply any relevant fixes from there into this PR.